Content.Client/UserInterface/ActionsUI.cs

```diff
@@ -0,0 +1,353 @@
+        private readonly ActionSlot[] _slots;
+
+        private readonly GridContainer _slotContainer;
+
+        private readonly TextureButton _lockButton;
+        private readonly TextureButton _settingsButton;
+        private readonly TextureButton _previousHotbarButton;
+        private readonly Label _loadoutNumber;
+        private readonly TextureButton _nextHotbarButton;
+        private readonly IClyde _clyde;
+        private readonly Texture _lockTexture;
+        private readonly Texture _unlockTexture;
+
+        private readonly TextureRect _dragShadow;
+        private readonly DragDropHelper<ActionSlot> _dragDropHelper;
+        public bool IsDragging => _dragDropHelper.IsDragging;
+        /// <summary>
+        /// Whether the bar is currently locked by the user. This is intended to prevent drag / drop
+        /// and right click clearing slots. Anything else is still doable.
+        /// </summary>
+        public bool Locked { get; private set; }
+
+        /// <summary>
+        /// All the action slots in order.
+        /// </summary>
+        public IEnumerable<ActionSlot> Slots => _slots;
+
+        /// <param name="onShowTooltip">OnShowTooltip handler to assign to each ActionSlot</param>
+        /// <param name="onHideTooltip">OnHideTooltip handler to assign to each ActionSlot</param>
+        /// <param name="onPressAction">OnPressed handler to assign to each action slot</param>
+        /// <param name="onDragDropAction">invoked when dragging and dropping an action from
+        /// one slot to another.</param>
+        /// <param name="onMouseEnteredAction">OnMouseEntered handler to assign to each action slot</param>
+        /// <param name="onMouseExitedAction">OnMouseExited handler to assign to each action slot</param>
+        /// <param name="onNextHotbarPressed">invoked when pressing the next hotbar button</param>
+        /// <param name="onPreviousHotbarPressed">invoked when pressing the previous hotbar button</param>
+        /// <param name="onSettingsButtonPressed">invoked when pressing the settings button</param>
+        public ActionsUI(EventHandler onShowTooltip, EventHandler onHideTooltip, Action<BaseButton.ButtonEventArgs> onPressAction,
+            Action<ActionSlotDragDropEventArgs> onDragDropAction, Action<GUIMouseHoverEventArgs> onMouseEnteredAction,
+            Action<GUIMouseHoverEventArgs> onMouseExitedAction,
+            Action<BaseButton.ButtonEventArgs> onNextHotbarPressed, Action<BaseButton.ButtonEventArgs> onPreviousHotbarPressed,
+            Action<BaseButton.ButtonEventArgs> onSettingsButtonPressed)
+        {
+            LayoutContainer.SetGrowHorizontal(this, LayoutContainer.GrowDirection.End);
+            LayoutContainer.SetAnchorAndMarginPreset(this, LayoutContainer.LayoutPreset.TopLeft, margin: 10);
+            LayoutContainer.SetMarginTop(this, 100);
+
+            _clyde = IoCManager.Resolve<IClyde>();
```

Comment:
Just couldn't figure out how to make that approach work, especially not without a "how much space do I have to work with" capability. I did try though. By "resize events" I assume you mean overriding "Resized()"?

If I'm not allowed to look at window size, that leaves me with control size. Resized() doesn't fire unless the control size is being changed by layoutcontainer. So in order to ensure that happens when the window is resized, I need to set a top and bottom anchor using layoutcontainer. Then what happens is, the control expands as it should as I increase window height. I can look at control.Height to see how much height I have to work with, and appropriately choose a single or 2 column layout.

But then if I shrink the height of the window, the control stays "stuck" at its previous height. The min size is calculated based on the contents of my control, which is currently dominated by the height of the single-column grid container inside it. So the control stops receiving Resized() calls and never knows when to switch back to the 2 column layout. Because layout container sees that the min height already overflows the "anchor-defined" height, so decides not to change the height. And my control can't know that it needs to switch to a 2 column layout (thus reducing the min height) because it doesn't have any sense of "how much space do I have to work with"...it just sees that the control size is currently the height of the single-column grid. It doesn't know "oh looks like we're getting squished".

So I don't see a way to accomplish this with the current UI system.

Reply:
You could override `CalculateMinimumSize` to return something like 0 to always allow the control to shrink.

Reply:
lol yep works perfectly. In my defense this seems like just as much of a hack as the other approach. But actually it makes sense, I just set the minimum height to the minimum height of the 2 column layout.

Reply:
Anecdotally, this would also be fixed by a WPF-style `Measure()` that takes in the available size. But this is a good solution either way.